django/contrib/admin/options.py

```diff
@@ -1902,43 +1960,64 @@ def get_formset(self, request, obj=None, **kwargs):
+            the constructor. It's a bit dirty but I don't know how to clean it
+            easily...
+            """
+            class DeleteProtectedModelForm(base_model_form):
+                def __init__(self, *args, **kwargs):
+                    super(DeleteProtectedModelForm, self).__init__(*args, **kwargs)
+                    if not can_change and not self.instance._state.adding:
+                        self.fields = {}
+                    if not can_add and self.instance._state.adding:
+                        self.fields = {}
+
+                def hand_clean_DELETE(self):
+                    """
+                    We don't validate the 'DELETE' field itself because on
+                    templates it's not rendered using the field information, but
+                    just using a generic "deletion_field" of the InlineModelAdmin.
+                    """
+                    if self.cleaned_data.get(DELETION_FIELD_NAME, False):
+                        using = router.db_for_write(self._meta.model)
+                        collector = NestedObjects(using=using)
+                        if self.instance._state.adding:
+                            return
+                        collector.collect([self.instance])
+                        if collector.protected:
+                            objs = []
+                            for p in collector.protected:
+                                objs.append(
+                                    # Translators: Model verbose name and instance representation,
+                                    # suitable to be an item in a list.
+                                    _('%(class_name)s %(instance)s') % {
+                                        'class_name': p._meta.verbose_name,
+                                        'instance': p}
+                                )
+                            params = {'class_name': self._meta.model._meta.verbose_name,
+                                      'instance': self.instance,
+                                      'related_objects': get_text_list(objs, _('and'))}
+                            msg = _("Deleting %(class_name)s %(instance)s would require "
+                                    "deleting the following protected related objects: "
+                                    "%(related_objects)s")
+                            raise ValidationError(msg, code='deleting_protected', params=params)
+
+                def is_valid(self):
+                    result = super().is_valid()
+                    self.hand_clean_DELETE()
+                    return result
+
+            return DeleteProtectedModelForm
+
+        can_change = True
+        can_add = True
+        if request:
+            can_change = self.has_change_permission(request, obj)
+            can_add = self.has_add_permission(request)
+
+        defaults['form'] = get_model_form(can_add, can_change)
```

Comment:
I must admin I'm not sure I completely understand what's happening here. As I said, I didn't do the initial implementation.

As far as I get it, despite it's name, the DeleteProtectedModelForm class is not only about deletion, but simply the base_model_form used for inlines. It seems to be needed to validate the DELETE field.

Then, about defaulting to true if there's no request, it's a good question. I'm not sure in which contexts it can happen that there's no request. I guess it's only outside of the regular admin context ? Or can it even happen ?